Add Terminal History via OSC Sequence Shell Injection

Fangoling opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description and review
the requirements below.

Contributors guide: https://github.com/theia-ide/theia/blob/master/CONTRIBUTING.md

Note: Security vulnerabilities should not be disclosed on GitHub, through a PR or any
other means. See SECURITY.md at the root of this repository, to learn how to report
vulnerabilities.
-->

#### What it does

<!-- Include relevant issues and describe how they are addressed. -->
This PR introduces **Terminal Command History** support to the Theia Terminal extension.

Each terminal widget now maintains an internal command history buffer that records discrete *command blocks*:
```
commandHistory: TerminalBlock[]

TerminalBlock {
    command: string;
    output: string;
}
```

To support this, the terminal now exposes two new events:
- **onTerminalCommandStart**
- **onTerminalPromptShow**

These events allow extensions to reliably detect when a prompt appears and when a command begins.

Additionally, this PR adds:
- **visual dividers between command blocks**, and  
- a **new user preference** to toggle these dividers (disabled by default).

**Issue**
The xterm.js terminal emulator does **not** expose APIs for retrieving past commands or outputs.  
Extensions such as *AI Terminal* currently extract a fixed number of buffer lines, which leads to issues:

- When the last terminal command is longer than the fixed number of lines, not the full context can be covered
- The model could focus on the wrong commands when more than one command is in the extracted output

**Solution**
We adopt the approach used in the [new JetBrains terminal architecture](https://blog.jetbrains.com/idea/2025/04/jetbrains-terminal-a-new-architecture/) by injecting **OSC 133 sequences** at the shell level.
These sequences indicate when prompts and commands begin and allow the terminal to reconstruct accurate command blocks.

```
\e]133;prompt_started\a
<terminal prompt>
\e]133;prompt_finished\a
\e]133;command_started;<terminal command>\a
<terminal command>
<terminal output>
\e]133;prompt_started\a
```

The terminal registers an OSC 133 handler and responds to different payloads to build the history.  
Because shells do not emit a “command finished” marker, the next appearing prompt implicitly finalizes the previous command block.

**Shell injection**
Shell integration must occur *inside the user’s shell environment*.  
To avoid overwriting user config, we follow the JetBrains strategy and load our integration scripts as wrappers:

**For zsh:**
- We provide stand-in versions of:
  - `.zlogin`, `.zprofile`, `.zshenv`, `.zshrc`
- Each stand-in sources the original file afterward
- Integration is activated via environment variables (e.g., `ZDOTDIR`)

**For bash:**
- The integration script is passed via startup arguments
- Only `.bashrc` requires sourcing

A new **ShellIntegrationInjector**:
- Detects the shell
- Applies the correct injection strategy
- Falls back gracefully when unsupported shells are used

**Package static resources**
- add a new entry to the Wepack Config generator that uses the `CopyWebpackPlugin` to copy the shell integration scripts into the lib folder of the packaged application

**Task Terminal Support**
Task terminals execute non-interactive commands (`bash -c "command"`) and cannot use shell integration hooks.
For these terminals, we inject OSC 133 sequences directly in the `TaskTerminalProcess`:
- `command_started` is emitted when the task begins execution
- `prompt_started` is emitted when the task exits or is killed
This approach is shell-agnostic and works consistently across all platforms (bash, zsh, cmd, PowerShell).
Unlike interactive terminals that rely on shell hooks, task terminals use direct OSC injection since they
execute commands in non-interactive mode where `.bashrc` and integration scripts are not sourced.



**Modified files**
The shell integration scripts are based on scripts from the [JetBrains IntelliJ community repository](https://github.com/JetBrains/intellij-community/tree/e49ad444f19179779ae7ab89c5dbab205ddf98dd/plugins/terminal/resources/shell-integrations), which were adapted and modified to fit our use case.

You can find them in `/packages/terminal/src/node/shell-integrations/`

`zsh/zsh-integration.zsh:`
- Adapted directory variable: `JETBRAINS_INTELLIJ_ZSH_DIR` -> `THEIA_ZSH_DIR`.
- Removed IntelliJ session bootstrap (JEDITERM_SOURCE and `_INTELLIJ_FORCE_SET_*` / `_INTELLIJ_FORCE_PREPEND_*` handling, plus the one-time `precmd_functions` hook injection).
- Removed legacy command block support handling.

`zsh/command-block-support.zsh`
- Renamed functions/vars from `__jetbrains_intellij_*` to `__theia_*`.
- Removed gating/early-returns for `INTELLIJ_TERMINAL_COMMAND_BLOCKS_REWORKED` and `P9K_VERSION` (integration always loads).
- Switched OSC protocol payloads from JetBrains `OSC 1341;...` to Theia `OSC 133;...` and changed `command_started` format to send only the encoded command (no `command=...` key).
- Dropped `command_finished` + `current_directory` reporting and removed alias reporting (`aliases_received`).

`zsh/zdotdir/`
.zlogin, .zprofile, .zshenv, .zshrc, source-original.zsh
- Rebranded variables from `JETBRAINS_INTELLIJ_*` to `THEIA_*`.
- Removed optional debug logging controlled by `JETBRAINS_INTELLIJ_TERMINAL_DEBUG_LOG_LEVEL`.

`bash/bash-integration.bash`
- Rebranded internals from JetBrains/IntelliJ to Theia (`JETBRAINS_INTELLIJ_BASH_DIR` -> `THEIA_BASH_DIR`, `__jetbrains_intellij_restore_posix_flag` -> `__theia_restore_posix_flag`).
- Removed IntelliJ-provided environment injection (`_INTELLIJ_FORCE_SET_*` / `_INTELLIJ_FORCE_PREPEND_*` processing), plus optional sourcing of `JEDITERM_USER_RCFILE` and `JEDITERM_SOURCE` (+ args).
- Dropped keybinding setup for Ctrl-left/right word movement.
- Dropped IntelliJ command-history hook (`__INTELLIJ_COMMAND_HISTFILE__` + EXIT trap).
- Only sources `command-block-support.bash` (no `command-block-support-reworked.bash`), without readability checks.

`bash/command-block-support.bash`
- Rebranded functions/vars from `__jetbrains_intellij_*` to `__theia_*`.
- Removed the feature gate (`INTELLIJ_TERMINAL_COMMAND_BLOCKS_REWORKED`) so the integration always loads.
- Switched OSC protocol from JetBrains `OSC 1341;...` to Theia `OSC 133;...` and changed `command_started` payload to emit only the encoded command (no `command=` key); `command_finished` no longer reports exit code/current directory.
- Improved command capture for `command_started` by reading the last history entry (fallback to `BASH_COMMAND`).


<!-- Explain how a reviewer can reproduce a bug, test new functionality or verify performance improvements. -->
#### How to test

I added debug logging to `initializeOSC133Support.ts` in the `terminal-widget-impl.ts` file.

1. Start the application with the `--log-level=debug` flag to enable `console.debug` output
2. Open a terminal
3. Open the developer tools console and enable: 
- verbose logging in chrome
- debug logging in firefox
5. Execute any command in the terminal.
6. Inspect the browser developer console.

Expected behavior:
- The full reconstructed command history of the current terminal is printed.
- The executed command and its complete output appear as a well-formed command block.

#### Follow-ups

<!-- Please list potential follow-up work, including known issues, possible future work, identified technical debt, and potentially introduced technical debt. If the PR introduces technical debt, specify the reason why this is acceptable. Please create tickets and link them here. Please use the label "technical debt" for new issues when it applies. -->

- **Z-index behavior:** Command block separators are rendered above the AI terminal widget.  
  This is also how other terminal decorations behave (e.g., selected text).

- the powerlevel10k theme for zsh does not work in combination with this feature, [this issue](https://youtrack.jetbrains.com/issue/IJPL-178955) is know to the jetbrains intellij team 

#### Breaking changes

- [ ] This PR introduces breaking changes and requires careful review. If yes, the breaking changes section in the [changelog](https://github.com/eclipse-theia/theia/blob/master/CHANGELOG.md) has been updated.

#### Attribution

<!-- If the changelog entry for this change should contain an attribution at the end (e.g. Contributed on behalf of x) add it in this section -->
Contributed under the supervision of @JonasHelming as part of the TUM Bachelor Thesis project "Enhancing Terminal Usability in Modern IDEs through AI-Assisted Interaction".

#### Review checklist

- [x] As an author, I have thoroughly tested my changes and carefully followed [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#requesting-a-review)
- [x] User-facing text is internationalized using the `nls` service (for details, please see the [Internationalization/Localization section](https://github.com/theia-ide/theia/blob/master/doc/coding-guidelines.md#internationalizationlocalization) in the Coding Guidelines)

#### Reminder for reviewers

- As a reviewer, I agree to behave in accordance with [the review guidelines](https://github.com/theia-ide/theia/blob/master/doc/pull-requests.md#reviewing)

sdirix commented on packages/terminal/src/node/shell-integrations/zsh/command-block-support.zsh:
```diff
@@ -0,0 +1,92 @@
+# and limitations under the License.
+#
+# Source:
+# https://github.com/JetBrains/intellij-community/blob/master/plugins/terminal/resources/shell-integrations/zsh/command-block-support-reworked.zsh
```
Permalink

sdirix commented on examples/browser/package.json:
```diff
@@ -94,14 +94,16 @@
     "build:production": "theiaext build && npm run -s bundle:production",
     "bundle": "npm run rebuild && theia build --mode development",
     "bundle:production": "npm run rebuild && theia build --mode production",
+    "postbuild": "copyfiles -u 6 '../../packages/terminal/src/node/shell-integrations/**' './lib/backend/' && copyfiles -u 6 '../../packages/terminal/src/node/shell-integrations/**/.*' './lib/backend/'",
```
We should not add a `postbuild` step here as it will only work for the Browser Example application.

If this is a step all adopters need to perform too, then we should integrate it into the generated webpack configuration instead.

sdirix commented on packages/terminal/src/node/shell-integrations/bash/bash-integration.bash:
```diff
@@ -0,0 +1,79 @@
+# https://github.com/JetBrains/intellij-community/blob/master/plugins/terminal/resources/shell-integrations/bash/bash-integration.bash
+# *****************************************************************************
+
+echo 'test'
```
Should be removed?
```suggestion
```

Fangoling replied:
yes

sgraband commented on packages/terminal/src/node/shell-integration-injector.ts:
```diff
@@ -0,0 +1,64 @@
+        const shellExecutable = options.shell ?? ShellProcess.getShellExecutablePath();
+        const shellType = guessShellTypeFromExecutable(shellExecutable);
+        if (shellType === GeneralShellType.Bash) {
+            // strips the login flag if present to avoid conflicts with --rcfile
```
Is this missing, or am i missing how this is done?

sgraband commented on packages/terminal/src/node/shell-integration-injector.ts:
```diff
@@ -0,0 +1,64 @@
+
+export class ShellIntegrationInjector {
+
+    static bashFlag = '--rcfile';
```
Could use a better name in my opinion. WDYT about bashInteractiveFlag?

Fangoling replied:
I renamed it to BASH_RCFILE_FLAG, I would not name it interactive flag, as it could be confused with the`--i` flag.

sgraband commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -216,6 +233,9 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         this.initializeLinkHover();
 
+        this.showCommandSeparator = this.preferences['terminal.integrated.enableCommandSeparator'] ?? false;
+        this.initializeOSC133Support();
```
Its fine to do this check this when the Terminal is created, but imho we also need a listener on the preference and update this value then according to the preference. Otherwise the setting only has effect when the user creates a new terminal. This makes it hard for the user to decide which option they like more.

sgraband commented on packages/debug/src/browser/debug-session.tsx:
```diff
@@ -521,6 +522,11 @@ export class DebugSession implements CompositeTreeElement {
+                await waitForEvent(terminal.onOutput, 1000);
+            } catch (error) {
+                console.warn(`Terminal did not emit output in time, using it anyway: ${error}`);
+            }
```
What is the point of waiting for the output if we continue anyway? This could use a comment if there is a difference. Also where do the 1000 ms come from?

sgraband replied:
Nitpick: I believe we do not delay the creation of the terminal, right? We delay opening it.

sgraband commented on examples/browser/package.json:
```diff
@@ -94,14 +94,16 @@
     "build:production": "theiaext build && npm run -s bundle:production",
     "bundle": "npm run rebuild && theia build --mode development",
     "bundle:production": "npm run rebuild && theia build --mode production",
+    "postbuild": "copyfiles -u 6 '../../packages/terminal/src/node/shell-integrations/**' './lib/backend/' && copyfiles -u 6 '../../packages/terminal/src/node/shell-integrations/**/.*' './lib/backend/'",
```
Could you check if this would also be achieveable with the `copy-webpack-plugin` we are already using? Then we wouldnt need to introduce a new dependency.

Fangoling replied:
That worked. I removed the copyfiles postbuild step and added the copy step to the webpack generation.

sgraband reviewed (changes requested):
Thank you for the contribution @Fangoling! This is already in a great state and works well (in the browser example)! I left some inline comments and questions. Could you take a look at those?

Note: I haven't reviewed the OSC13 parsing and the bash/zsh files.

Fangoling commented on packages/terminal/src/node/shell-integrations/bash/bash-integration.bash:
```diff
@@ -0,0 +1,79 @@
+#!/bin/bash
+# *****************************************************************************
+# Copyright (C) 2000-2025 JetBrains s.r.o.
+# Modifications (C) 2025 Fangxing Liu
```
Sure. Should I also reference Eclipse Source instead of my name here?

Fangoling commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -325,6 +346,13 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
         this.toDispose.push(this.searchBox);
+
+        this.term.onKey(({ domEvent }) => {
+            if (domEvent.key === 'Enter') {
```
The domEvent.key Enter captures all onKey events that include the Enter key.

The special case where Alt+Enter creates a new line without executing the command results in only one entry in the command history. Mapping commands to output in this context is not currently supported.

Fangoling commented on packages/ai-terminal/src/browser/ai-terminal-contribution.ts:
```diff
@@ -185,6 +185,14 @@ class AiTerminalChatWidget {
 
     protected getRecentTerminalCommands(): string[] {
+        const commandHistory = this.terminalWidget.commandHistory;
+        console.log('Full command history:', commandHistory);
```
I removed them here and added logs on `debug` level directly to the osc handler function.

sgraband reviewed (commented):
Thank you for the changes @Fangoling! Just one small inline comments, and some comments on other threads from my side.

@sdirix Do you want to take another look?

Fangoling commented:
> I will review at the latest next week.

I added the preference for terminal command history that you requested:
- add experimental preference setting for enabling terminal history
- add experimental flag to terminal divider and state that it depends on terminal history being enabled
- disable usage/injection of shell config when terminal history is disabled
- disable command history in ai-terminal when terminal history is disabled
- disable rendering of terminal dividers when terminal history is disabled
- changing the preference does not impact already created terminals

I also added the missing support for task terminals, since these do not use an interactive shell I am not able to reuse the shell injector.
- I am injecting the terminal sequences by extending the `TaskTerminalProcess` in `packages/task/src/node/process/process-task-runner.ts`.
- Currently the injection of osc sequences into the output is not disabled when disabling the command history.

sdirix commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -165,11 +166,27 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+        return this._commandHistory;
+    }
+
+    private enableCommandHistory: boolean;
```
The `enableCommandHistory` field is never read (without also setting it at the same time). It can be removed

sdirix commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -165,11 +166,27 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+    }
+
+    private enableCommandHistory: boolean;
+    private showCommandSeparator: boolean;
```
The `showCommandSeparator` is read one time in `initializeOSC133Support` but it's set AFTER `initializeOSC133Support() is called.
I think we can just read from preference in `initializeOSC133Support` and remove the field.

sdirix commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -204,7 +221,8 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
             fastScrollSensitivity: this.preferences['terminal.integrated.fastScrollSensitivity'],
-            theme: this.themeService.theme
+            theme: this.themeService.theme,
+            allowProposedApi: true
```
We should document why we need proposed api. Should we set `false` in case we don't have command history enabled?

sdirix commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -216,6 +234,12 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+        this.enableCommandHistory = this.preferences.get('terminal.integrated.enableCommandHistory') ?? false;
+        if (this.enableCommandHistory) {
+            this.initializeOSC133Support();
+            this.showCommandSeparator = this.preferences.get('terminal.integrated.enableCommandSeparator') ?? false;
```
same here
```suggestion
            this.showCommandSeparator = this.preferences.get('terminal.integrated.enableCommandSeparator', false);
```

sdirix commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -344,6 +376,10 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+        this.enableCommandHistory = this.preferences.get('terminal.integrated.enableCommandHistory') ?? false;
+        this.showCommandSeparator = this.enableCommandHistory
+            ? this.preferences.get('terminal.integrated.enableCommandSeparator') ?? false
+            : false;
```
```suggestion
        this.showCommandSeparator = enableCommandHistory && this.preferences.get('terminal.integrated.enableCommandSeparator', false);
```

sdirix commented on packages/terminal/src/browser/terminal-widget-impl.ts:
```diff
@@ -1009,4 +1048,73 @@ export class TerminalWidgetImpl extends TerminalWidget implements StatefulWidget
+                    output: this.sanitizeCommandOutput(this.commandOutputBuffer)
+                };
+                console.debug('Current command history:', this._commandHistory);
+                console.debug('Terminal command result captured:', terminalBlock);
```
We have a named logger available here

Fangoling commented on packages/terminal/src/browser/base/terminal-widget.ts:
```diff
@@ -181,6 +186,21 @@ export abstract class TerminalWidget extends BaseWidget {
+    /**
+     * Event which fires when terminal prompt is shown.
+     */
+    abstract onTerminalPromptShown: Event<void>;
```
Yes. It meant for clients that need this information.

Fangoling commented on packages/terminal/src/browser/style/terminal.css:
```diff
@@ -30,3 +30,10 @@
+  border-top: var(--theia-border-width) solid var(--theia-sideBarSectionHeader-border);
+  left: 0 !important;
+  right: 0 !important;
+  width: 99% !important;
```
Changed it to 100% as it looks more consistent if margin on both sides is the same.

Fangoling commented on packages/terminal/src/node/terminal-backend-module.ts:
```diff
@@ -61,7 +62,15 @@ export default new ContainerModule(bind => {
+
+            // inject shell integration scripts and env vars only if the terminal command history is enabled
+            const preferences = ctx.container.get<TerminalPreferences>(TerminalPreferences);
+            const enableCommandHistory = preferences['terminal.integrated.enableCommandHistory'] ?? false;
```
I modified
- `packages/terminal/src/node/shell-process.ts`
- `packages/terminal/src/common/shell-terminal-protocol.ts`

and added the parameter `enableShellIntegration` to `shellProcessOptions` and `IShellTerminalServerOptions`

This allows us to pass in the preference when calling `shellTerminalServer.create()` in `TerminalWidgetImpl` on Client side and access it in `terminal-backend.module`

Fangoling commented:
> Thanks for the update!
> 
> Please check my comments.
> 
> I tested it on my Ubuntu machine, it did work for me with bash but it did not work for zsh:
> 
> <img alt="Image" width="1358" height="260" src="https://private-user-images.githubusercontent.com/8998368/536284752-880c5707-4138-4763-ad69-0894fb9fa1d2.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Njg1MjE5MzEsIm5iZiI6MTc2ODUyMTYzMSwicGF0aCI6Ii84OTk4MzY4LzUzNjI4NDc1Mi04ODBjNTcwNy00MTM4LTQ3NjMtYWQ2OS0wODk0ZmI5ZmExZDIucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDExNiUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAxMTZUMDAwMDMxWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9MDk4OTM4YjgxZmZlNTA3YzJmY2UyY2ZmNjU3MmEwZWU2ZDZmMDlkOTNiNGY4NzI3MDdhNmY1MTg3OTk3OGE1MiZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.773MQRy-wCSM4JnpvyKfwi14xgxTU7lzifh-cz0WBsQ">

Thats really funny. The script that JetBrains provides that we are using for this functionality does not work when you are using the [Powerlevel10k](https://github.com/romkatv/powerlevel10k) theme. Here is an issue related to this https://youtrack.jetbrains.com/issue/IJPL-178955.

The [original scripts](https://github.com/JetBrains/intellij-community/blob/master/plugins/terminal/resources/shell-integrations/zsh/command-block-support-reworked.zsh) disables support completly and does not source the integration script to begin with. I removed it back then as I thought that it was just an edge case, but apparently a lot of people use this theme. 

Until that is fixed we could find our own solution or add this as a follow up issue.

sdirix commented on packages/terminal/src/node/shell-integration-injector.ts:
```diff
@@ -0,0 +1,80 @@
+    }
+
+    private static getShellIntegrationPath(relativePath: string): string {
+        return path.join(__dirname, this.INTEGRATION_ROOT_DIR, relativePath);
```
`__dirname` at runtime will be the compiled `lib/node/` directory. The shell scripts live in `src/node/shell-integrations/` and are only available in the output if copied by CopyPlugin (or the Electron build performs an equivalent step). If the copy doesn't happen (e.g., Electron app without the webpack plugin, or a custom build), this will silently fail with `ENOENT`. Consider adding an existence check or a clear error message."

Fangoling commented on packages/terminal/src/node/shell-integrations/zsh/command-block-support.zsh:
```diff
@@ -0,0 +1,107 @@
+add-zsh-hook preexec __theia_command_preexec
+add-zsh-hook precmd __theia_command_precmd
+
+if [[ -n "$THEIA_USER_ZSHRC" && -f "$THEIA_USER_ZSHRC" ]]; then
```
It is also not present in the linked command-block implementation by jetbrains.
I will removed it, since we do not use this property.

sdirix commented on dev-packages/application-manager/src/generator/webpack-generator.ts:
```diff
@@ -95,6 +95,12 @@ const plugins = [
+                // copy shell integration scripts
+                from: path.join(resolvePackagePath('@theia/terminal', __dirname), '..', 'src', 'node', 'shell-integrations'),
+                to: path.resolve(__dirname, 'lib', 'backend', 'shell-integrations')
+            }`)}
```
We had discussions about an adopter guide in the past but in fact we never added it. So ignore my comment 👍

sdirix commented:
@jfaltermeier Can you check whether the Theia IDE packaging needs to be adapted to integrate the shell injection scripts?
